### PR TITLE
Added JSON support to user preferences

### PR DESCRIPTION
**Pull request by migurski**
I’m unsure about this, because I haven’t seen an example with more than one preference.

Test plan:

```
GET /api/0.6/user/preferences?format=json

{"preference":{"k":"diary.default_language","v":"en"}}
```

```
GET /api/0.6/user/preferences?format=xml

<osm version="0.6" generator="OpenStreetMap server" copyright="OpenStreetMap and contributors" attribution="http://www.openstreetmap.org/copyright"license="http://opendatacommons.org/licenses/odbl/1-0/">
<preferences>
<preference k="diary.default_language" v="en"/>
</preferences>
</osm>
```

Closes https://github.com/openstreetmap/openstreetmap-website/issues/3160

### CODE REVIEW

**Review by pnorman — CHANGES_REQUESTED** [review]
> {"preference":{"k":"diary.default_language","v":"en"}}

I think `{"preference": {"diary.default_language": "en"}}` would be better, unless duplicate keys are possible.

**Comment by migurski** [comment]
Oh good call. Updated in fd75268:

```
GET /api/0.6/user/preferences?format=json

{"preference":{"diary.default_language":"en"}}
```

**Comment by tomhughes** [comment]
Is there a reason not to just drop the `preference` containers and do a simple dictionary of key value pairs in the JSON output?

The `preference` is needed in XML to avoid having arbitrary elements but a dictionary would be the natural approach for JSON I think.

File: app/views/api/user_preferences/index.json.jbuilder [hunk 1]
```diff
@@ -0,0 +1 @@
+json.partial! @user_preferences
```
Comment (gravitystorm) [inline app/views/api/user_preferences/index.json.jbuilder:1]: This should also contain the `root_elements`, similar to the XML response. This is already implemented for e.g. showing relations - see https://github.com/openstreetmap/openstreetmap-website/blob/master/app/views/api/relations/show.json.jbuilder

When those are included, it becomes clearer that there needs to be a root `preferences` (n.b. plural) key, to match the naming of the `<preferences>` element in the XML response. 

So something like this:

```
{
    "version": "0.6",
    "generator": "OpenStreetMap server",
    "copyright": "OpenStreetMap and contributors",
    "attribution": "http://www.openstreetmap.org/copyright",
    "license": "http://opendatacommons.org/licenses/odbl/1-0/",
    "preferences": {
        "foo": "bar",
        "another_pref": "something else",
        "more preferences": "go here"
    }
}
```

**Review by gravitystorm — CHANGES_REQUESTED** [review]
This PR also needs tests. Similarly to the previous JSON-response PRs, it doesn't need to duplicate every test of the XML output, but just enough to ensure that the routes exist, and that when json is requested a suitable json-formatted response is received.

Have a look at test/controllers/api/users_controller_test.rb and search for 'json' to see what I mean.

File: app/views/api/user_preferences/index.json.jbuilder (hunk 1 above)
Comment (gravitystorm) [reply app/views/api/user_preferences/index.json.jbuilder:1]: I should also say that the preferences are analogous to "tags" on nodes/ways/relations, in that they are already implemented in the nodes/way/relations responses so you can see how it's done there, and for json we already decided to drop the "k/v" stuff so as @pnorman mentioned it's fine to use object attributes directly for them here too.

**Comment by gravitystorm** [comment]
There is also the topic of the other methods (show, update, update_all, delete) and implementing a json API for each of them. But that opens a small can of worms, since those existing methods are terrible (e.g. plaintext responses to xml payloads, empty responses for updating a user_preference vs repeating the payload when updating a node) and so I'm happy if you want to back away slowly from that part of the discussion.

File: app/views/api/user_preferences/index.json.jbuilder (hunk 1 above)
Comment (migurski) [reply app/views/api/user_preferences/index.json.jbuilder:1]: Thanks, I’ll add these! Do you know of a way to see multiple preferences? So far I’ve only been able to view the editor language, and I’d like to have additional key/value pairs.

**Comment by migurski** [comment]
I may back away slowly from that part for now.

File: app/views/api/user_preferences/index.json.jbuilder (hunk 1 above)
Comment (pnorman) [reply app/views/api/user_preferences/index.json.jbuilder:1]: You'll have to upload preferences using [the api](https://wiki.openstreetmap.org/wiki/API_v0.6#Preferences_of_the_logged-in_user)

**Comment by migurski** [comment]
Updated in a95931ea3:

```
GET /api/0.6/user/preferences?format=json

{
    "attribution": "http://www.openstreetmap.org/copyright",
    "copyright": "OpenStreetMap and contributors",
    "generator": "OpenStreetMap server",
    "license": "http://opendatacommons.org/licenses/odbl/1-0/",
    "preferences": {
        "coffee": "cream, no sugar",
        "diary.default_language": "en"
    },
    "version": "0.6"
}
```

```
GET /api/0.6/user/preferences?format=xml

<osm version="0.6" generator="OpenStreetMap server" copyright="OpenStreetMap and contributors" attribution="http://www.openstreetmap.org/copyright" license="http://opendatacommons.org/licenses/odbl/1-0/">
  <preferences>
    <preference k="diary.default_language" v="en"/>
    <preference k="coffee" v="cream, no sugar"/>
  </preferences>
</osm>
```

**Comment by migurski** [comment]
Tests coming soon, while I learn how tests work in Rails.

**Comment by migurski** [comment]
Tests done, all feedback addressed. I think this is ready to merge.

**Comment by gravitystorm** [comment]
> Tests done, all feedback addressed. I think this is ready to merge.

Great, thanks! I also spotted a couple of minor issues with the tests, but nothing major. You can see my changes at 6bb18436a31bc192c20f3df5caf15ac5888471b5 - adding a test for the routes, adding an assertion to ensure the output is in the correct format, and then tweaking a comment that drifted out of date.

Thanks again!

**Pull request merged**